// gnim-hooks@0.1.0/src/gtk4/useStyle.ts
import { Accessor, createComputed, createState, onCleanup } from "gnim"
import Gdk from "gi://Gdk?version=4.0"
import Gtk from "gi://Gtk?version=4.0"
import { $ } from "../reactive.js"

type ClassValue = string | number | null | boolean | undefined

export function cls<Args extends Array<$<ClassValue>>>(...args: Args): Accessor<string> {
    return createComputed(
        args.map((a) => $(a)),
        (...values) => values.filter((v) => typeof v === "string" && v !== "").join(" "),
    )
}

function applyCss(stylesheet: string): () => void {
    const display = Gdk.Display.get_default()!

    const provider = new Gtk.CssProvider()
    provider.load_from_data(stylesheet, -1)

    let id = provider.connect("parsing-error", (_, section, error) => {
        const start = section.get_start_location()
        const end = section.get_end_location()

        printerr("css error:", error.message)
        printerr("in section", section.to_string())
        printerr(
            "at location",
            `start: ${start.lines}:${start.chars}`,
            `end: ${end.lines}:${end.chars}`,
        )
    })

    Gtk.StyleContext.add_provider_for_display(display, provider, Gtk.STYLE_PROVIDER_PRIORITY_USER)

    return () => {
        if (id > 0) {
            provider.disconnect(id)
            Gtk.StyleContext.remove_provider_for_display(display, provider)
            id = 0
        }
    }
}

export function css(css: TemplateStringsArray, ...values: $<string>[]): () => void {
    let cleanup: (() => void) | null = null

    const stylesheet = createComputed(
        values.map((v) => $(v)),
        (...values) => css.flatMap((str, i) => str + `${values[i] ?? ""}`).join(""),
    )

    function sync() {
        cleanup?.()
        cleanup = applyCss(stylesheet.get())
    }

    sync()
    const unsub = stylesheet.subscribe(sync)

    return () => {
        cleanup?.()
        unsub()
    }
}

const cache = new Map<string, { cleanup: () => void; counter: number }>()

function fnv1aHash(str: string) {
    let hash = 0x811c9dc5
    for (let i = 0; i < str.length; i++) {
        hash ^= str.charCodeAt(i)
        hash = (hash * 0x01000193) >>> 0
    }
    return hash
}

function styleSheet(selector: string, style: Style): string[] {
    const { "&": nested = {}, ...props } = style

    const attrs = Object.entries(props)
        .map(([attr, v]) => `${attr}: ${v};`)
        .join("")

    const nestedStyles = Object.entries(nested).flatMap(([pseudo, props]) =>
        styleSheet(selector + pseudo, props),
    )

    return [`${selector} { ${attrs} }`, ...nestedStyles]
}

function createClassName(style: Style): [string, () => void] {
    const id = "s" + fnv1aHash(JSON.stringify(style)).toString(36)
    let css = cache.get(id)

    if (!css) {
        const cleanup = applyCss(styleSheet(`.${id}`, style).join(" "))

        css = { cleanup, counter: 1 }
        cache.set(id, css)
    } else {
        css.counter += 1
    }

    return [
        id,
        () => {
            css.counter -= 1
            setTimeout(() => {
                if (css.counter === 0 && cache.has(id)) {
                    cache.delete(id)
                    css.cleanup()
                }
            }, 5_000)
        },
    ]
}

export function useStyle(props: Style): string
export function useStyle(props: Accessor<Style>): Accessor<string>

export function useStyle(props: $<Style>): $<string> {
    if (props instanceof Accessor) {
        let [c, cleanup] = createClassName(props.get())
        const [name, setName] = createState(c)

        const unsub = props.subscribe(() => {
            cleanup()
            const [newClass, newCleanup] = createClassName(props.get())
            setName((c = newClass))
            cleanup = newCleanup
        })

        onCleanup(() => {
            cleanup()
            unsub()
        })

        return name
    }

    const [c, cleanup] = createClassName(props)
    onCleanup(cleanup)
    return c
}

// TODO:
// - support media queries & keyframes
// - support camelCase attributes to not need quoting
export interface Style extends Partial<Record<CssProperty, string | string[]>> {
    "&"?: {
        [key: string]: Style
    }
}

// https://docs.gtk.org/gtk4/css-properties.html#gtk-css-properties
export type CssProperty =
    | `--${string}`
    | "all"
    | "color"
    | "opacity"
    | "filter"
    | "font-family"
    | "font-size"
    | "font-style"
    | "font-variant"
    | "font-weight"
    | "font-stretch"
    | "font-kerning"
    | "font-variant-ligatures"
    | "font-variant-position"
    | "font-variant-caps"
    | "font-variant-numeric"
    | "font-variant-alternates"
    | "font-variant-east-asian"
    | "font-feature-settings"
    | "font-variation-settings"
    | "-gtk-dpi"
    | "font"
    | "caret-color"
    | "-gtk-secondary-caret-color"
    | "letter-spacing"
    | "text-transform"
    | "line-height"
    | "text-decoration-line"
    | "text-decoration-color"
    | "text-decoration-style"
    | "text-shadow"
    | "text-decoration"
    | "-gtk-icon-source"
    | "-gtk-icon-size"
    | "-gtk-icon-style"
    | "-gtk-icon-transform"
    | "-gtk-icon-palette"
    | "-gtk-icon-shadow"
    | "-gtk-icon-filter"
    | "transform"
    | "transform-origin"
    | "min-width"
    | "min-height"
    | "margin-top"
    | "margin-right"
    | "margin-bottom"
    | "margin-left"
    | "padding-top"
    | "padding-right"
    | "padding-bottom"
    | "padding-left"
    | "margin"
    | "padding"
    | "border-top-width"
    | "border-right-width"
    | "border-bottom-width"
    | "border-left-width"
    | "border-top-style"
    | "border-right-style"
    | "border-bottom-style"
    | "border-left-style"
    | "border-top-right-radius"
    | "border-bottom-right-radius"
    | "border-bottom-left-radius"
    | "border-top-left-radius"
    | "border-top-color"
    | "border-right-color"
    | "border-bottom-color"
    | "border-left-color"
    | "border-image-source"
    | "border-image-repeat"
    | "border-image-slice"
    | "border-image-width"
    | "border-width"
    | "border-style"
    | "border-color"
    | "border-top"
    | "border-right"
    | "border-bottom"
    | "border-left"
    | "border"
    | "border-radius"
    | "border-image"
    | "outline-style"
    | "outline-width"
    | "outline-color"
    | "outline-offset"
    | "outline"
    | "background-color"
    | "background-clip"
    | "background-origin"
    | "background-size"
    | "background-position"
    | "background-repeat"
    | "background-image"
    | "box-shadow"
    | "background-blend-mode"
    | "background"
    | "transition-property"
    | "transition-duration"
    | "transition-timing-function"
    | "transition-delay"
    | "transition"
    | "animation-name"
    | "animation-duration"
    | "animation-timing-function"
    | "animation-iteration-count"
    | "animation-direction"
    | "animation-play-state"
    | "animation-delay"
    | "animation-fill-mode"
    | "animation"
    | "border-spacing"
